Added chunk view API

Flo0 opened the pull request:
This exposes some functionality of the underlying ChunkTrackingView.

When adding certain objects to chunks, it is often desirable to have actual knowledge about who is currently seeing this chunk, in order to selectively send informations to those viewers.

This can be useful for holograms and similar functionality.

Flo0 committed:
Added chunk view API

Cryptite commented on patches/server/1055-Added-API-implementation.patch:
```diff
@@ -0,0 +1,61 @@
++        net.minecraft.world.level.ChunkPos chunkPos = new net.minecraft.world.level.ChunkPos(chunk.getX(), chunk.getZ());
++        return this.getHandle().getChunkTrackingView().contains(chunkPos);
++    }
++    // Paper end
```
You can use `this.getHandle().chunkLoader.sentChunks.contains(chunk.getChunkKey());` to access the already-existing list of ChunkKeys stored by the server, rather than constructing a new ChunkPos.

`sentChunks` may need to be made public however, that's what I did on my fork (for nms access reasons)

Flo0 committed:
API for checking sent chunks

Flo0 committed:
API for checking sent chunks

Vrganj commented on patches/server/1055-API-for-checking-sent-chunks.patch:
```diff
@@ -0,0 +1,58 @@
++    }
++
++    @Override
++    public boolean isChunkSent(org.bukkit.Chunk chunk) {
```
Why not use the chunk key here instead of Chunk, if you opted for using keys above

Cryptite replied:
Or at least provide an extra overload which accepts a key.

lynxplay commented on patches/server/1055-API-for-checking-sent-chunks.patch:
```diff
@@ -0,0 +1,63 @@
+ import org.bukkit.entity.Player;
+ import java.lang.invoke.VarHandle;
+ import java.util.ArrayDeque;
++import java.util.Set;
```
Inline the import please.

lynxplay commented on patches/api/0469-API-for-checking-sent-chunks.patch:
```diff
@@ -0,0 +1,42 @@
++     * @param chunkKey the chunk key to check
++     * @return true if the player has been sent the chunk, false otherwise
++     */
++    boolean isChunkSent(java.lang.Long chunkKey);
```
What is the point of a boxed long here?
The implementation requires a primitive long in the first place. Unless I am mistakes it should be using a fastutil long set.

Flo0 replied:
No use here, changed to primitive.

lynxplay commented on patches/api/0469-API-for-checking-sent-chunks.patch:
```diff
@@ -0,0 +1,42 @@
++     *
++     * @return an immutable set of chunk keys
++     */
++    java.util.Set<java.lang.Long> getSentChunkKeys();
```
Can be marked with @org.jetbrains.annotations.Unmodifiable.
Additionally, the NotNull annotation.

Flo0 replied:
Unmodifiable has a retention policy of class only i think

lynxplay replied:
Correct. It isn't needed at runtime, but Intellij can infer more about the returned collection via said annotation.

lynxplay commented on patches/api/0469-API-for-checking-sent-chunks.patch:
```diff
@@ -0,0 +1,42 @@
++     * @param chunk the chunk to check
++     * @return true if the player has been sent the chunk, false otherwise
++     */
++    boolean isChunkSent(org.bukkit.Chunk chunk);
```
Missing NotNull annotation.
This can also just be a default method tbh.

lynxplay commented on patches/api/0469-API-for-checking-sent-chunks.patch:
```diff
@@ -0,0 +1,42 @@
++    /**
++     * Checks if the player has been sent a specific chunk.
++     *
++     * @param chunkKey the chunk key to check
```
I'd link to Chunk#getChunkKey here via a `@see` or something as the concept of a chunk key, or well, its creation is at least somewhat explained there.

lynxplay committed:
Rebase and changes

lynxplay committed:
Avoid long boxing

lynxplay committed:
Mark as experimental for now

Given the argumentation between yielding a view and yielding a copy of
the sent chunk set, these methods are for now marked as experimental to
gather feedback from consumers and potentially change their behaviour
down the line.
